lib/logstash/filters/kubernetes_metadata.rb

```diff
@@ -0,0 +1,288 @@
+      @logger.debug("kubernetes file info got: #{metadata}")
+
+      pod = file_metadata['pod']
+      namespace = file_metadata['namespace']
+      name = file_metadata['container_name']
+
+      return unless pod and namespace and name
+
+      if data = get_kubernetes(namespace, pod)
+        metadata = file_metadata.merge(data)
+        set_log_formats(metadata)
+        lookup_cache[path] = metadata
+      end
+      @logger.debug("metadata within lookup_cache[path]: #{metadata}")
+    end
+
+
+    @logger.debug("metadata after unless lookup_cache[path] is: #{metadata}")
+    @logger.debug("config after unless lookup_cache[path] is: #{config}")
+
+    event.set(@target, metadata)
+    return filter_matched(event)
+  end
+
+  def set_log_formats(metadata)
+    begin
+      #return if metadata['annotations'].empty?
+
+      format = {
+        'stderr' => @default_log_format,
+        'stdout' => @default_log_format
+      }
+      a = metadata['annotations']
+      n = metadata['container_name']
+
+      # check for log-format-<stream>-<name>, log-format-<name>, log-format-<stream>, log-format
+      # in annotations
+      %w{ stderr stdout }.each do |t|
+        [ "log-format-#{t}-#{n}", "log-format-#{n}", "log-format-#{t}", "log-format" ].each do |k|
+          if v = a[k]
+            format[t] = v
+            break
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata => #{metadata}")
+
+    rescue => e
+      @logger.warn("Error setting log format: #{e}")
+    end
+  end
+
+  # based on https://github.com/vaijab/logstash-filter-kubernetes/blob/master/lib/logstash/filters/kubernetes.rb
+  def get_file_info(path)
+    parts = path.split(File::SEPARATOR).last.gsub(/.log$/, '').split('_')
+    if parts.length != 3 || parts[2].start_with?('POD-')
+      return nil
+    end
+    kubernetes = {}
+    kubernetes['replication_controller'] = parts[0].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['pod'] = parts[0]
+    kubernetes['namespace'] = parts[1]
+    kubernetes['container_name'] = parts[2].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['container_id'] = parts[2].split('-').last
+    return kubernetes
+  end
+
+  def sanatize_keys(data)
+    return {} unless data
+
+    parsed_data = {}
+    data.each do |k,v|
+      new_key = k.gsub(/\.|,/, '_')
+        .gsub(/\//, '-')
+      parsed_data[new_key] = v
+    end
+
+    return parsed_data
+  end
+
+  def get_kubernetes(namespace, pod)
+
+    begin
+      @logger.debug("Attempting to query the Kubernetes API")
+
+      url = [ @api, 'api/v1/namespaces', namespace, 'pods', pod ].join("/")
+
+      rest_opts = {
+        verify_ssl: false
+      }
+
+      if @auth
+        if @auth['basic']
+          @logger.debug("Found basic auth for Kubernetes API")
+
+          basic_user = @auth['basic']['user']
+          basic_pass = @auth['basic']['pass']
```

Comment:
Can you add extra validation here for these settings? It would be easy for a user to typo 'user' and 'pass' (and 'basic' and 'bearer', etc) and they won't get any meaningful feedback from the configuration validator about this.